scarlet/source.py

```diff
@@ -395,6 +396,104 @@ def __init__(self, frame, sky_coord, observations):
         return self.frame.psf.__call__(*parameters, bbox=self.bbox)[0]
 
+class StarletSource(FunctionComponent):
+    """Source intialized with starlet coefficients.
+
+    Sources are initialized with the SED of the center pixel,
+    and the morphologies are initialised as ExtendedSources
+    and transformed into starlet coefficients.
+    """
+
+    def __init__(
+        self,
+        frame,
+        sky_coord,
+        observations,
+        obs_idx=0,
+        thresh=1.0,
+        shifting=False,
+    ):
+        """Extended source intialized to match a set of observations
+
+        Parameters
+        ----------
+        frame: `~scarlet.Frame`
+            The frame of the model
+        sky_coord: tuple
+            Center of the source
+        observations: instance or list of `~scarlet.observation.Observation`
+            Observation(s) to initialize this source.
+        obs_idx: int
+            Index of the observation in `observations` to
+            initialize the morphology.
+        thresh: `float`
+            Multiple of the backround RMS used as a
+            flux cutoff for morphology initialization.
+        shifting: `bool`
+            Whether or not a subpixel shift is added as optimization parameter
+        """
+        center = np.array(frame.get_pixel(sky_coord), dtype="float")
+        self.pixel_center = tuple(np.round(center).astype("int"))
+
+        # initialize SED from sky_coord
+        try:
+            iter(observations)
+        except TypeError:
+            observations = [observations]
+
+        # initialize from observation
+        sed, image_morph, bbox = init_extended_source(
+            sky_coord,
+            frame,
+            observations,
+            obs_idx=obs_idx,
+            thresh=thresh,
+            symmetric=True,
+            monotonic=True,
+        )
+
+        sed = Parameter(
+            sed,
+            name="sed",
+            step=partial(relative_step, factor=1e-2),
+            constraint=PositivityConstraint(),
+        )
+
+        noise = mad_wavelet(observations[0].images) * \
+                np.sum(observations[0]._diff_kernels.image**2, axis = (-2,-1))**0.5
+        # Threshold in units of noise
+        thresh = 5 * np.sqrt(np.sum((sed*noise) ** 2))
+
+        # Starlet transform of morphologies (n1,n2) with 4 dimensions: (1,lvl,n1,n2), lvl = wavelet scales
+        self.transform = Starlet(image_morph)
+        #The starlet transform is the model
+        morph = self.transform.coefficients*0
+        # wavelet-scale norm
+        starlet_norm = self.transform.norm
+        #One threshold per wavelet scale: thresh*norm
+        thresh_array = np.zeros(morph.shape) + thresh
+        thresh_array = thresh_array * np.array([starlet_norm])[..., np.newaxis, np.newaxis]
+        # We don't threshold the last scale
+        thresh_array[:,-1,:,:] = 0
+
+        morph_constraint = ConstraintChain(*[L0Constraint(thresh_array), PositivityConstraint()])
```

Comment:
Hmm: are starlet coefficients strictly positive for positive sources?

Reply:
Yes, this generation of wavelets was specially designed for this. But it doesn't mean that a sparsest representation cannot be achieved that includes negative coefficients. The actual way to do this is to do sub-iterations to find the point at the intersection of the set of positive-valued real images and the set of images with a starlet decomposition in the desired L1(0) ball.  
It's not super clean to do it this way.

Reply:
If strictly non-negative coefficients result in strictly non-negative images, our approach also yields a feasible solution. Is this condition satisfied by starlets?

Reply:
Yup! By design of this specific familly (second generation starlets)
<img width="863" alt="Screen Shot 2020-05-13 at 4 50 55 PM" src="https://user-images.githubusercontent.com/16084926/81863942-f1c33c80-9539-11ea-9a13-c9112e329285.png">
How cool would it be to to be able to type and execute small codes directly inside comment bubbles?!

Reply:
Excellent!